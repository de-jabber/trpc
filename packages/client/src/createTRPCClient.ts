/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { Unsubscribable } from '@trpc/server/observable';
import type {
  AnyProcedure,
  AnyRouter,
  inferClientTypes,
  inferProcedureInput,
  InferrableClientTypes,
  inferTransformedProcedureOutput,
  ProcedureType,
  RouterRecord,
} from '@trpc/server/unstable-core-do-not-import';
import {
  createFlatProxy,
  createRecursiveProxy,
} from '@trpc/server/unstable-core-do-not-import';
import type { CreateTRPCClientOptions } from './createTRPCUntypedClient';
import type { TRPCSubscriptionObserver } from './internals/TRPCUntypedClient';
import { TRPCUntypedClient } from './internals/TRPCUntypedClient';
import type { TRPCProcedureOptions } from './internals/types';
import type { TRPCClientError } from './TRPCClientError';

/**
 * @public
 **/
export type inferRouterClient<TRouter extends AnyRouter> =
  DecoratedProcedureRecord<TRouter, TRouter['_def']['record']>;

type ResolverDef = {
  input: any;
  output: any;
  transformer: boolean;
  errorShape: any;
};

type coerceAsyncGeneratorToIterable<T> =
  T extends AsyncGenerator<infer $T, infer $Return, infer $Next>
    ? AsyncIterable<$T, $Return, $Next>
    : T;

/** @internal */
export type Resolver<TDef extends ResolverDef> = (
  input: TDef['input'],
  opts?: TRPCProcedureOptions,
) => Promise<coerceAsyncGeneratorToIterable<TDef['output']>>;

type SubscriptionResolver<TDef extends ResolverDef> = (
  input: TDef['input'],
  opts: Partial<
    TRPCSubscriptionObserver<TDef['output'], TRPCClientError<TDef>>
  > &
    TRPCProcedureOptions,
) => Unsubscribable;

type DecorateProcedure<
  TType extends ProcedureType,
  TDef extends ResolverDef,
> = TType extends 'query'
  ? {
      query: Resolver<TDef>;
    }
  : TType extends 'mutation'
    ? {
        mutate: Resolver<TDef>;
      }
    : TType extends 'subscription'
      ? {
          subscribe: SubscriptionResolver<TDef>;
        }
      : never;

/**
 * @internal
 */
type DecoratedProcedureRecord<
  TRoot extends InferrableClientTypes,
  TRecord extends RouterRecord,
> = {
  [TKey in keyof TRecord]: TRecord[TKey] extends infer $Value
    ? $Value extends AnyProcedure
      ? DecorateProcedure<
          $Value['_def']['type'],
          {
            input: inferProcedureInput<$Value>;
            output: inferTransformedProcedureOutput<
              inferClientTypes<TRoot>,
              $Value
            >;
            errorShape: inferClientTypes<TRoot>['errorShape'];
            transformer: inferClientTypes<TRoot>['transformer'];
          }
        >
      : $Value extends RouterRecord
        ? DecoratedProcedureRecord<TRoot, $Value>
        : never
    : never;
};

const clientCallTypeMap: Record<
  keyof DecorateProcedure<any, any>,
  ProcedureType
> = {
  query: 'query',
  mutate: 'mutation',
  subscribe: 'subscription',
};

/** @internal */
export const clientCallTypeToProcedureType = (
  clientCallType: string,
): ProcedureType => {
  return clientCallTypeMap[clientCallType as keyof typeof clientCallTypeMap];
};

/**
 * @internal
 */
export function createTRPCClientProxy<TRouter extends AnyRouter>(
  client: TRPCUntypedClient<TRouter>,
): inferRouterClient<TRouter> {
  const proxy = createRecursiveProxy<inferRouterClient<TRouter>>(
    ({ path, args }) => {
      const pathCopy = [...path];
      const procedureType = clientCallTypeToProcedureType(pathCopy.pop()!);

      const fullPath = pathCopy.join('.');

      return (client as any)[procedureType](fullPath, ...args);
    },
  );
  return createFlatProxy<inferRouterClient<TRouter>>((key) => {
    if (key === '__untypedClient') {
      return client;
    }
    return proxy[key];
  });
}

export function createTRPCClient<TRouter extends AnyRouter>(
  opts: CreateTRPCClientOptions<TRouter>,
): inferRouterClient<TRouter> {
  const client = new TRPCUntypedClient(opts);
  const proxy = createTRPCClientProxy<TRouter>(client);
  return proxy;
}

/**
 * Get an untyped client from a proxy client
 * @internal
 */
export function getUntypedClient<TRouter extends AnyRouter>(
  client: inferRouterClient<TRouter>,
): TRPCUntypedClient<TRouter> {
  return (client as any).__untypedClient;
}
