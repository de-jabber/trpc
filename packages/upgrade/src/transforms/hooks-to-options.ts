import type {
  API,
  ArrowFunctionExpression,
  ASTPath,
  FileInfo,
  FunctionDeclaration,
  Options,
} from 'jscodeshift';

interface TransformOptions extends Options {
  trpcFile?: string;
  trpcImportName?: string;
}

const hookToOptions = {
  useQuery: { lib: '@tanstack/react-query', fn: 'queryOptions' },
  useSuspenseQuery: { lib: '@tanstack/react-query', fn: 'queryOptions' },
  useInfiniteQuery: {
    lib: '@tanstack/react-query',
    fn: 'infiniteQueryOptions',
  },
  useSuspenseInfiniteQuery: {
    lib: '@tanstack/react-query',
    fn: 'infiniteQueryOptions',
  },
  useMutation: { lib: '@tanstack/react-query', fn: 'mutationOptions' },
  useSubscription: {
    lib: '@trpc/tanstack-react-query',
    fn: 'subscriptionOptions',
  },
} as const;

export default function transform(
  file: FileInfo,
  api: API,
  options: TransformOptions,
) {
  const { trpcFile, trpcImportName } = options;

  const j = api.jscodeshift;
  const root = j(file.source);
  let dirtyFlag = false;

  function addUseTRPCToFunction(
    path: ASTPath<FunctionDeclaration | ArrowFunctionExpression>,
  ) {
    const body = j.BlockStatement.check(path.node.body)
      ? path.node.body.body
      : [path.node.body];

    body.unshift(
      j.variableDeclaration('const', [
        j.variableDeclarator(
          j.identifier(trpcImportName),
          j.callExpression(j.identifier('useTRPC'), []),
        ),
      ]),
    );
    dirtyFlag = true;
  }

  function ensureImported(lib: string, specifier: string) {
    // Upsert RQ import
    if (
      root
        .find(j.ImportDeclaration, {
          source: { value: lib },
        })
        .find(j.ImportSpecifier, { imported: { name: specifier } })
        .size() === 0
    ) {
      root
        .find(j.ImportDeclaration)
        .at(-1)
        .insertAfter(
          j.importDeclaration(
            [j.importSpecifier(j.identifier(specifier))],
            j.literal(lib),
          ),
        );
      dirtyFlag = true;
    }
  }

  // Update trpc import to useTRPC - skip if it's `trpc.createClient`
  root
    .find(j.ImportDeclaration, { source: { value: trpcFile } })
    .forEach((path) => {
      const trpcSpecifier = path.node.specifiers?.find((specifier) => {
        return (
          j.ImportSpecifier.check(specifier) &&
          specifier.imported.name === trpcImportName
        );
      });

      if (trpcSpecifier) {
        path.node.specifiers = path.node.specifiers?.map((specifier) => {
          if (
            j.ImportSpecifier.check(specifier) &&
            specifier.imported.name === trpcImportName
          ) {
            return j.importSpecifier(j.identifier('useTRPC'));
          }
          return specifier;
        });
        dirtyFlag = true;
      }
    });

  root.find(j.FunctionDeclaration).forEach((path) => {
    const usesTrpc =
      j(path).find(j.Identifier, { name: trpcImportName }).size() > 0;

    if (usesTrpc) {
      addUseTRPCToFunction(path);
    }
  });

  root.find(j.ArrowFunctionExpression).forEach((path) => {
    const usesTrpc =
      j(path)
        .find(j.MemberExpression, {
          object: { name: trpcImportName },
          property: { name: (name: string) => name !== 'createClient' },
        })
        .size() > 0;
    if (usesTrpc) {
      addUseTRPCToFunction(path);
    }
  });

  // REplace proxy-hooks with useX(options())
  for (const [hook, { fn, lib }] of Object.entries(hookToOptions)) {
    root
      .find(j.CallExpression, {
        callee: {
          property: { name: hook },
          object: { object: { object: { name: 'trpc' } } },
        },
      })
      .forEach((path) => {
        const memberExpr = path.node.callee;
        memberExpr.property.name = fn;

        const useQueryFunction = j.callExpression(j.identifier(hook), [
          path.node,
        ]);
        j(path).replaceWith(useQueryFunction);

        ensureImported(lib, hook);
        dirtyFlag = true;
      });
  }

  // Remove suspense query destructuring
  root.find(j.VariableDeclaration).forEach((path) => {
    const declarator = j.VariableDeclarator.check(path.node.declarations[0])
      ? path.node.declarations[0]
      : null;

    if (
      !j.CallExpression.check(declarator?.init) ||
      !j.Identifier.check(declarator.init.callee) ||
      declarator.init.callee.name !== 'useSuspenseQuery'
    ) {
      return;
    }

    const tuple = j.ArrayPattern.check(declarator?.id) ? declarator.id : null;
    const dataName = j.Identifier.check(tuple?.elements?.[0])
      ? tuple.elements[0].name
      : null;
    const queryName = j.Identifier.check(tuple?.elements?.[1])
      ? tuple.elements[1].name
      : null;

    if (declarator && dataName && queryName) {
      declarator.id = j.identifier(queryName);
      j(path).insertAfter(
        j.variableDeclaration('const', [
          j.variableDeclarator(
            j.identifier(dataName),
            j.memberExpression(j.identifier(queryName), j.identifier('data')),
          ),
        ]),
      );
    }
  });

  return dirtyFlag ? root.toSource() : undefined;
}

export const parser = 'tsx';

// https://go.codemod.com/ddX54TM
