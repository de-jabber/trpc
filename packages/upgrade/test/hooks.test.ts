import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { applyTransform } from 'jscodeshift/src/testUtils';
import { format, resolveConfig } from 'prettier';
import { expect, it } from 'vitest';
import * as transform from '../src/transforms/hooks-to-options';

const formatFile = async (path: string, source: string) => {
  const prettierConfig = await resolveConfig(path);
  return format(source, {
    ...prettierConfig,
    parser: 'typescript',
  });
};

const testFixture = async (file: string) => {
  const source = readFileSync(file, 'utf-8');

  const transformed = applyTransform(transform, {}, { source });
  const formatted = await formatFile(file, transformed);

  expect(formatted).toMatchSnapshot();
};

const literal = './fixtures';
const fixturesDir = new URL(literal, import.meta.url).pathname;

const fixtures = readdirSync(fixturesDir);
it.each(fixtures)(
  '%s',
  async (file) => await testFixture(join(fixturesDir, file)),
);
