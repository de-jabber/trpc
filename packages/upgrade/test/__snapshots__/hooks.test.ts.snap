// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic.ts 1`] = `
"import { useQuery } from '@tanstack/react-query';
import { trpc } from './trpc';

export function Component() {
  useQuery(trpc.post.list.queryOptions());
  useQuery(trpc.post.queryOptions());
  useQuery(trpc.a.b.c.d.queryOptions());

  useQuery(trpc.post.byId.queryOptions({ id: 1 }));
  useQuery(trpc.num.queryOptions(1));

  useQuery(trpc.post.list.queryOptions(undefined, { staleTime: 1000 }));
  useQuery(trpc.post.byId.queryOptions({ id: 1 }, { staleTime: 1000 }));
}
"
`;

exports[`existing-rq.ts 1`] = `""`;

exports[`existing-rq-mix.ts 1`] = `
"import { useInfiniteQuery, useMutation, useQuery } from '@tanstack/react-query';
import { trpc } from './trpc';

export function Component() {
  useQuery({ queryKey: ['a'], queryFn: () => 1 });
  useMutation({ mutationFn: () => 1 });

  useQuery(trpc.post.byId.queryOptions({ id: 1 }));
  useQuery(trpc.num.queryOptions(1));
  useInfiniteQuery(trpc.post.list.infiniteQueryOptions());
}
"
`;

exports[`hooks-basic-query.ts 1`] = `
"import { useQuery } from '@tanstack/react-query';
import { trpc } from './trpc';

export function Component() {
  useQuery(trpc.post.list.queryOptions());
  useQuery(trpc.post.queryOptions());
  useQuery(trpc.a.b.c.d.queryOptions());

  useQuery(trpc.post.byId.queryOptions({ id: 1 }));
  useQuery(trpc.num.queryOptions(1));

  useQuery(trpc.post.list.queryOptions(undefined, { staleTime: 1000 }));
  useQuery(trpc.post.byId.queryOptions({ id: 1 }, { staleTime: 1000 }));
}
"
`;

exports[`suspense-destructuring.ts 1`] = `
"import { useSuspenseQuery } from '@tanstack/react-query';
import { trpc } from './trpc';

export function Component() {
  const query = useSuspenseQuery(trpc.post.list.queryOptions());

  const data = query.data;

  const [a, b] = [1, 2];
}
"
`;

exports[`with-invalidate.ts 1`] = `
"import { useMutation, useQuery } from '@tanstack/react-query';
import { trpc } from './trpc';

export function Component() {
  const { data } = useQuery(trpc.post.list.queryOptions());
  const queryClient = useQueryClient();

  const mutation = useMutation(
    trpc.post.create.mutationOptions({
      onSettled: () => queryClient.invalidateQueries(trpc.post.queryFilter()),
    }),
  );
}
"
`;
